api-authentication/server/index.js

```diff
@@ -0,0 +1,104 @@
+require('dotenv/config');
+const pg = require('pg');
+const argon2 = require('argon2'); // eslint-disable-line
+const express = require('express');
+const jwt = require('jsonwebtoken'); // eslint-disable-line
+const ClientError = require('./client-error');
+const errorMiddleware = require('./error-middleware');
+
+const db = new pg.Pool({ // eslint-disable-line
+  connectionString: process.env.DATABASE_URL,
+  ssl: {
+    rejectUnauthorized: false
+  }
+});
+
+const app = express();
+
+const jsonMiddleware = express.json();
+
+app.use(jsonMiddleware);
+
+app.post('/api/auth/sign-up', (req, res, next) => {
+  const { username, password } = req.body;
+  if (!username || !password) {
+    throw new ClientError(400, 'username and password are required fields');
+  }
+  argon2
+    .hash(password)
+    .then(hashedPassword => {
+      const sql = `
+        insert into "users" ("username", "hashedPassword")
+        values ($1, $2)
+        returning "userId", "username", "createdAt"
+      `;
+      const params = [username, hashedPassword];
+      return db.query(sql, params);
+    })
+    .then(result => {
+      const [user] = result.rows;
+      res.status(201).json(user);
+    })
+    .catch(err => next(err));
+});
+
+app.post('/api/auth/sign-in', (req, res, next) => {
+  const { username, password } = req.body;
+  if (!username || !password) {
+    throw new ClientError(401, 'invalid login');
+  }
+
+  /* your code starts here */
+const sql = `
+  SELECT "username",
+         "hashedPassword"
+          from "users"
+          where "username" = ($1)
+
+  `
+  const params = [username]
+  db.query(sql, params)
+.then(result => {
+  const [user] = result.rows
+  if(!user) {
+    throw new ClientError(404, 'Invalid Login')
+  }
+  console.log(result)
+  const {userId, hashedPassword} = user;
+  argon2
+  .verify(hashedPassword, password)
+    .then(isMacthing => {
```

Comment:
Hey Brian, the only update here is that if it does not match you want to send back a 401 client error.